### PR TITLE
Stop ignoring compilation errors in UTs

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/CFG/Sonar/SonarCfgSerializerTest.cs [hunk 1]
```diff
@@ -382,9 +382,9 @@ public void Range()
 ");
         }
 
-        protected static IControlFlowGraph GetCfgForMethod(string code, string methodName)
+        private static IControlFlowGraph GetCfgForMethod(string code, string methodName)
         {
-            (var method, var semanticModel) = TestHelper.Compile(code).GetMethod(methodName);
+            (var method, var semanticModel) = TestHelper.CompileIgnoreErrors(code).GetMethod(methodName);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/CFG/Sonar/SonarCfgSerializerTest.cs:387]: Accepting fact that SonarCfgSerializer uses a lot of snippets with errors

File: analyzers/tests/SonarAnalyzer.UnitTest/CFG/Sonar/SonarControlFlowGraphTest.cs [hunk 2]
```diff
@@ -5066,7 +5066,7 @@ public void Bar()
         internal static MethodDeclarationSyntax CompileWithMethodBody(string input, string methodName, out SemanticModel semanticModel, ParseOptions parseOptions = null)
         {
             MethodDeclarationSyntax methodDeclaration;
-            (methodDeclaration, semanticModel) = TestHelper.Compile(input).GetMethod(methodName);
+            (methodDeclaration, semanticModel) = TestHelper.CompileIgnoreErrors(input).GetMethod(methodName);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/CFG/Sonar/SonarControlFlowGraphTest.cs:5069]: Accepting snippets with errors for sonar CFG

File: analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs [hunk 3]
```diff
@@ -563,17 +586,16 @@ End If
...
                 throw new ArgumentException("Supplied language is not C# neither VB.Net", nameof(language));
             }
 
-            var (syntaxTree, semanticModel) = TestHelper.Compile(text);
+            var (syntaxTree, semanticModel) = TestHelper.Compile(text, language == AnalyzerLanguage.CSharp);
 
             return language == AnalyzerLanguage.CSharp
                 ? new Metrics.CSharp.CSharpMetrics(syntaxTree, semanticModel)
-                : new Metrics.VisualBasic.VisualBasicMetrics(VisualBasicSyntaxTree.ParseText(text), semanticModel);
+                : new Metrics.VisualBasic.VisualBasicMetrics(syntaxTree, semanticModel);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs:598]: All VB test cases were compiled as C# code and then we computed metrics totally wrong

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs [hunk 4]
```diff
@@ -217,7 +217,7 @@ public void Main()
         Undefined(() => 45);
     }
 }";
-            var (tree, semanticModel) = TestHelper.Compile(code);
+            var (tree, semanticModel) = TestHelper.CompileIgnoreErrors(code);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:220]: Compile errors are expected in this test

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs [hunk 5]
```diff
@@ -230,7 +230,7 @@ public void Main()
         [DataRow(@"{ () => => =>")]
         public void CreateCfg_InvalidSyntax_ReturnsCfg(string code)
         {
-            var (tree, semanticModel) = TestHelper.Compile(code);
+            var (tree, semanticModel) = TestHelper.CompileIgnoreErrors(code);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/SyntaxNodeExtensionsTest.cs:233]: Compile errors are expected in this test

File: analyzers/tests/SonarAnalyzer.UnitTest/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.cs [hunk 6]
```diff
@@ -777,22 +779,6 @@ public void ProcessVariableInForeach_Declared_LiveIn_LiveOut()
...
-foreach(i in new int[] {1, 2, 3})
-{
-    Method(i, intParameter);
-}";
-            var context = CreateContextCS(code);
-            context.Validate(context.Cfg.Blocks[1], null, new LiveIn("intParameter"), new LiveOut("intParameter"));
-            context.Validate(context.Cfg.Blocks[2], null, new LiveIn("intParameter"), new LiveOut("intParameter"));
-            context.Validate("Method(i, intParameter);", new LiveIn("intParameter"), new LiveOut("intParameter"));
-            context.ValidateExit();
-        }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.cs:794]: Foreach variable cannot be reused => this UT doesn't make sense. There's another one to test normal `foreach`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/TestHelper.cs [hunk 7]
```diff
@@ -47,18 +47,16 @@ internal static class TestHelper
     <OutPath>{2}</OutPath>
 </SonarProjectConfig>";
 
-        public static (SyntaxTree, SemanticModel) Compile(string classDeclaration, bool isCSharp = true, params MetadataReference[] additionalReferences)
+        public static (SyntaxTree, SemanticModel) CompileIgnoreErrors(string snippet, bool isCSharp = true, params MetadataReference[] additionalReferences) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/TestHelper.cs:50]: This file has the core change of this PR

File: analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs [hunk 8]
```diff
@@ -33,35 +33,35 @@ public class MetricsTest
         [TestMethod]
         public void Lines()
         {
-            Lines(AnalyzerLanguage.CSharp, "").Should().Be(1);
+            Lines(AnalyzerLanguage.CSharp, string.Empty).Should().Be(1);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs:36]: Almost everything was wrong in this file (see comment at the EOF)

File: analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs [hunk 9]
```diff
@@ -189,306 +189,327 @@ public void CommentsWithHeaders()
...
-            Statements(AnalyzerLanguage.VisualBasic, "Class MyClass \n Sub MyMethod() \n Console.WriteLine() \n End Sub \n End Class").Should().Be(1);
-            Statements(AnalyzerLanguage.VisualBasic, "Class MyClass \n Sub MyMethod() \n Foo:  \n End Sub \n End Class").Should().Be(1);
-            Statements(AnalyzerLanguage.VisualBasic, "Class MyClass \n Sub MyMethod() \n GoTo Foo \n End Sub \n End Class").Should().Be(1);
-            Statements(AnalyzerLanguage.VisualBasic, "Class MyClass \n Sub MyMethod() \n Break \n End Sub \n End Class").Should().Be(1);
-            Statements(AnalyzerLanguage.VisualBasic, "Class MyClass \n Sub MyMethod() \n Continue \n End Sub \n End Class").Should().Be(1);
-            Statements(AnalyzerLanguage.VisualBasic, "Class MyClass \n Sub MyMethod() \n Throw \n End Sub \n End Class").Should().Be(1);
-            Statements(AnalyzerLanguage.VisualBasic, "Class MyClass \n Sub MyMethod() \n Yield 42 \n End Sub \n End Class").Should().Be(1);
-            Statements(AnalyzerLanguage.VisualBasic, "Class MyClass \n Sub MyMethod() \n yield Exit While \n End Sub \n End Class").Should().Be(1);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs:294]: Removed, there's no `yield` keyword in VB

File: analyzers/tests/SonarAnalyzer.UnitTest/Metrics/VisualBasicExecutableLinesMetricTest.cs [hunk 10]
```diff
@@ -28,296 +28,414 @@ public class VisualBasicExecutableLinesMetricTest
     {
         [TestMethod]
         public void AttributeList() =>
-            AssertLinesOfCode(@"
-<System.Runtime.InteropServices.DllImport(""user32.dll"")>
-<System.Serializable()>
-Sub SampleMethod()
-End Sub");
+            AssertLinesOfCode(
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Metrics/VisualBasicExecutableLinesMetricTest.cs:31]: Everything was wrong in this file

File: analyzers/tests/SonarAnalyzer.UnitTest/Metrics/CSharpExecutableLinesMetricTest.cs [hunk 11]
```diff
@@ -30,885 +29,874 @@ public class CSharpExecutableLinesMetricTest
         [TestMethod]
         public void No_Executable_Lines() =>
             AssertLinesOfCode(
-              @"using System;
-                using System.Linq;
+@"using System;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Metrics/CSharpExecutableLinesMetricTest.cs:32]: There's a lot of whitespace cleanup for readability. This file should be easier to review per commit

File: analyzers/src/SonarAnalyzer.VisualBasic/Metrics/VisualBasicExecutableLinesMetric.cs [hunk 12]
```diff
@@ -97,6 +97,7 @@ private bool FindExecutableLines(SyntaxNode node)
                     case SyntaxKind.WhileStatement:
 
                     case SyntaxKind.IfStatement:
+                    case SyntaxKind.ElseIfStatement:
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Metrics/VisualBasicExecutableLinesMetric.cs:100]: See this commit and the modified UT for more context: [`882a1e7` (#5082)](https://github.com/SonarSource/sonar-dotnet/pull/5082/commits/882a1e757a5a2eeba160d7696278cc6f4e489963)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SnippetCompiler.cs [hunk 13]
```diff
@@ -154,14 +154,15 @@ public IPropertySymbol GetPropertySymbol(string typeDotMethodName)
...
+        private static bool IsCompilationError(Diagnostic diagnostic) =>
+            diagnostic.Severity == DiagnosticSeverity.Error && (diagnostic.Id.StartsWith("CS") || diagnostic.Id.StartsWith("BC"));
 
         private static void DumpCompilationErrors(Compilation compilation)
         {
-            var diagnostics = compilation.GetDiagnostics();
-
             Console.WriteLine("Diagnostic errors:");
-            foreach (var d in diagnostics)
+            foreach (var d in compilation.GetDiagnostics().Where(IsCompilationError))
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SnippetCompiler.cs:165]: _[random thought, not actionable]_ when I see this change I realize (as a code reviewer) that having tests for the test framework would be nice in the context of this PR

File: analyzers/tests/SonarAnalyzer.UnitTest/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.LocalFunction.cs [hunk 14]
```diff
@@ -275,15 +275,15 @@ public void LocalFunctionInvocation_CrossReference_WhenAnalyzingLocalFunctionIts
 int LocalFunction()
 {
     var variable = 42;
-    if (boolCondition)
+    if (boolParameter)
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.LocalFunction.cs:278]: review note: `CreateContextCS` puts this snippet inside a function that has `boolParameter` parameter

File: analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs [hunk 15]
```diff
@@ -189,306 +189,327 @@ public void CommentsWithHeaders()
         [TestMethod]
         public void Classes()
         {
-            Classes(AnalyzerLanguage.CSharp, "").Should().Be(0);
-            Classes(AnalyzerLanguage.CSharp, "class MyClass {}").Should().Be(1);
+            Classes(AnalyzerLanguage.CSharp, string.Empty).Should().Be(0);
+            Classes(AnalyzerLanguage.CSharp, "class Sample {}").Should().Be(1);
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs:193]: [curiosity, not actionable]  why was this change needed?
if it was to simplify, then why
```
IMyInterface
MyEnum
MyRecord
```
haven't been updated, like in `Accessors()`?

File: analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs [hunk 16]
```diff
@@ -189,306 +189,327 @@ public void CommentsWithHeaders()
...
             Classes(AnalyzerLanguage.CSharp, "enum MyEnum {}").Should().Be(0);
-            Classes(AnalyzerLanguage.CSharp, "class MyClass1 {} namespace MyNamespace { class MyClass2 {} }").Should().Be(2);
+            Classes(AnalyzerLanguage.CSharp, "class Sample1 {} namespace MyNamespace { class Sample2 {} }").Should().Be(2);
 
-            Classes(AnalyzerLanguage.VisualBasic, "").Should().Be(0);
+            Classes(AnalyzerLanguage.VisualBasic, string.Empty).Should().Be(0);
             Classes(AnalyzerLanguage.VisualBasic, "Class M \n End Class").Should().Be(1);
             Classes(AnalyzerLanguage.VisualBasic, "Structure M \n End Structure").Should().Be(1);
-            Classes(AnalyzerLanguage.VisualBasic, "Enum M \n End Enum").Should().Be(0);
+            Classes(AnalyzerLanguage.VisualBasic, "Enum M \n None \n End Enum").Should().Be(0);
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs:208]: [curiosity, not actionable] why not use here-string?
```
@"
Enum M
  None
End Enum
"
```

File: analyzers/src/SonarAnalyzer.VisualBasic/Metrics/VisualBasicExecutableLinesMetric.cs (hunk 12 above)
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.VisualBasic/Metrics/VisualBasicExecutableLinesMetric.cs:100]: This should be in the release notes as it's a functional change

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM, it's great to see cleaned tech debt!

Before merging, please:
- cherry-pick the changes on `VisualBasicExecutableLinesMetric.cs` to a different branch & PR (that should go into the release notes of next release, as it is a functional change)
- undo the changes on `VisualBasicExecutableLinesMetric.cs` in this PR (this PR should not be in the release notes)

For my other questions, I only asked out of curiosity. This is my only request for this PR (because the functional change can be easily missed in a 700 LOC change PR :) )

File: analyzers/src/SonarAnalyzer.VisualBasic/Metrics/VisualBasicExecutableLinesMetric.cs (hunk 12 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.VisualBasic/Metrics/VisualBasicExecutableLinesMetric.cs:100]: See #5109

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SnippetCompiler.cs (hunk 13 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SnippetCompiler.cs:165]: It actually is actionable. We do have some UTs for some parts of the TestFramework. I can at least bootstrap the class to have at least something.

File: analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs (hunk 15 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Common/MetricsTest.cs:193]: VB.NET has a `MyClass` keyword that would require VB and C# naming to be different (more painful).

I also find `class MyClass` more difficult to read. `class Sample` clearly distinguishes between the structure (keyword) and data.
